### PR TITLE
[PIP-163][Txn]Add lowWaterMark check before appending entry to TB

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Master Issue: [#15423](https://github.com/apache/pulsar/issues/15423)
### Motivation & Modification
Details can be found at https://github.com/apache/pulsar/issues/15423.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 1]
```diff
@@ -240,6 +244,13 @@ public CompletableFuture<Void> checkIfTBRecoverCompletely(boolean isTxnEnabled)
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
         CompletableFuture<Position> completableFuture = new CompletableFuture<>();
+        if (lowWaterMarks.get(txnId.getMostSigBits()) != null
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:247]: Add a local variable for `lowWaterMarks.get(txnId.getMostSigBits())`

**Review by congbobo184 — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 2]
```diff
@@ -275,6 +286,7 @@ public CompletableFuture<TransactionBufferReader> openTransactionBufferReader(Tx
 
     @Override
     public CompletableFuture<Void> commitTxn(TxnID txnID, long lowWaterMark) {
+        lowWaterMarks.compute(txnID.getMostSigBits(), (k, v) -> lowWaterMark);
```
[the remark this answers is not in the document]
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:289]: Good suggestion, I think you are right.  The method `endTxnInTransactionBuffer` is asynchronous, we can not grantee the order of committed operation .

**Review by poorbarcode — APPROVED** [review]
Good work

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java [hunk 3]
```diff
@@ -240,6 +244,13 @@ public CompletableFuture<Void> checkIfTBRecoverCompletely(boolean isTxnEnabled)
     @Override
     public CompletableFuture<Position> appendBufferToTxn(TxnID txnId, long sequenceId, ByteBuf buffer) {
         CompletableFuture<Position> completableFuture = new CompletableFuture<>();
+        Long lowWaterMark = lowWaterMarks.get(txnId.getMostSigBits());
+        if (lowWaterMark != null && lowWaterMark >= txnId.getLeastSigBits()) {
+            completableFuture.completeExceptionally(new TransactionBufferException
+                    .TransactionNotFoundException("Transaction [" + txnId + "] has been ended. "
+                    + "Please use a new transaction to send message."));
+            return completableFuture;
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:252]: Please check here https://github.com/apache/pulsar/blob/4f53e8eedc8aca406f60ab4c1f7bc11577a6bf20/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java#L2977-L2984

The TransactionNotFoundException will be wrapped with ManagedLedgerException. It should be wrapped by PersistenceException directly?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java [hunk 4]
```diff
@@ -971,4 +972,21 @@ public void testPendingAckMarkDeletePosition() throws Exception {
                     Integer.parseInt(lastConfirmedEntry[1]) - 2);
         });
     }
+
+    @Test
+    public void testTBLowWaterMark() throws Exception {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:977]: Please also add an end to end test to make sure the client-side can handle the new exception well.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java (hunk 4 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java:977]: We already have end to end test at `TransactionLowWaterMarkTest`.

Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java]: I have move the test from `TransactionTest` to  `TransactionLowWaterMarkTest`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java (hunk 3 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TopicTransactionBuffer.java:251]: I have change exception from `TransactionNotFoundException` to `NotAllowedException`.
And the `NotAllowedException` is not only used to `Exceed maximum message size`, so i change  the `recoverNotAllowedError` to make the expression of the log clearer.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java (hunk 4 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/TransactionTest.java]: OK, the test is very much needed.

**Review by codelipenghui — APPROVED** [review]

**Pull request merged**